### PR TITLE
LABELING: add script to interface between LabelMe and ROS bags

### PR DESCRIPTION
**Pull request by kev-the-dev**
Add a script which takes a YAML configuration and can generate ROS bags with labeled image data or put images for ROS bags into a LabelMe installation. Can also generate a report on what percentage of images have been labeled. See argparse config at bottom for usage.

Has been tested on a LabelMe installation successfully with provided example config. From this yaml, images were put onto LabelMe. Once labeled, the same script was use to produce a new bag with all of the original data with label data added at correct times.

<small>#Langforded</small>

### CODE REVIEW

File: utils/mil_tools/nodes/label_me_bag.py
Comment (sentree) [inline utils/mil_tools/nodes/label_me_bag.py]: Looks like you got the Langford brand of spelling down as well.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (mattlangford) [inline utils/mil_tools/nodes/label_me_bag.py]: Woah cowboy, slow down. Mixing yer 4 and 2 spaces.

File: utils/mil_tools/nodes/label_me_bag_example.yaml
Comment (mattlangford) [inline utils/mil_tools/nodes/label_me_bag_example.yaml]: Can you support -1.0 or something for a `stop:` (to dump until the end of the bag).

File: utils/mil_tools/nodes/label_me_bag.py
Comment (mattlangford) [inline utils/mil_tools/nodes/label_me_bag.py]: If you have shorthand args like `-r` it may be good to make the full length arguments more descriptive, i.e.`--extract-labels` or `--report-labels`. It will be easier to remember when you have to call this script a month or two down the line (so you're not having to `-h` everytime you run the script). Also when you have scripts with 10 or so arguments, clear names help a lot.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (mattlangford) [inline utils/mil_tools/nodes/label_me_bag.py]: `fprint`?

File: utils/mil_tools/nodes/label_me_bag_example.yaml
Comment (mattlangford) [inline utils/mil_tools/nodes/label_me_bag_example.yaml]: Also, you should allow 0.0 `freq` to dump everything. (You'd have to check for zero in `_read_bag` else you'd divide by 0)

File: utils/mil_tools/nodes/label_me_bag.py
Comment (mattlangford) [inline utils/mil_tools/nodes/label_me_bag.py]: Maybe have this function print the intent of the script, i.e. "I will dump images from {topic} to {outdir} on {bagname} from {start} to {end} every {freq} seconds." for each entry in the yaml. I know you do this down in the bag function, but having it all print out at once may be more useful. That would let you verify everything will run smoothly and leave your computer and let it work (I don't know how long this takes to run but on multiple big bags it could be rough)

This would have been cool for the mission runner back in the good old days of Navigator.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (mattlangford) [inline utils/mil_tools/nodes/label_me_bag.py]: You do the percent function a lot, maybe break it out somewhere to clean up code?

**Review by mattlangford — APPROVED** [review]
I think this is a good first run. It may be difficult to get times for every bag you want to label though right? That would require going through all the bags you want images for and keeping track of start and end times.
This might be a good case for an addition to the rqt bag plugin that would let you play a bag and make selections there.

What do you think the intended workflow would be?

File: utils/mil_tools/nodes/label_me_bag.py
Comment (kev-the-dev) [reply utils/mil_tools/nodes/label_me_bag.py]: Great idea. I'll also add a dry run flag to describe what it going to happen based on YAML and not actually touch and files.

File: utils/mil_tools/nodes/label_me_bag_example.yaml
Comment (kev-the-dev) [reply utils/mil_tools/nodes/label_me_bag_example.yaml]: Would be useful. I'll probably just accept a missing start or stop parameter and assume 0.0 and END_OF_BAG respectively.

**Comment by kev-the-dev** [comment]
While it will be tedious to select segment times, I'm not sure of any other option (if a program could do it, there would be no point to our perception). For finding times for the path marker task, it wasn't too hard. I just pulled up rqt and toggled the bag with spacebar to pick some good times. Writing notes on what is in bags when we take them will help a lot.

**Comment by mattlangford** [comment]
Yeah, taking bags specific for the vision task is a good idea.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (sentree) [inline utils/mil_tools/nodes/label_me_bag.py]: I like that you're using private functions and variables here, but consider using double underscores at the beginning of the name. A single underscore implies private for other programmers and fits our convention, but double underscores will actually make it hard to access from outside the object. Since this is more similar to a truly private function or variable, it is what I use.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (sentree) [inline utils/mil_tools/nodes/label_me_bag.py]: I feel like python has a more elegant logging interface, but it's not a big deal.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (sentree) [inline utils/mil_tools/nodes/label_me_bag.py]: Do you intend to have a master YAML for all of the bags?

File: utils/mil_tools/nodes/label_me_bag.py
Comment (sentree) [inline utils/mil_tools/nodes/label_me_bag.py]: This should probably happen before merging.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (sentree) [inline utils/mil_tools/nodes/label_me_bag.py]: Can you return the percentage so this is usable by another program? I think I could have the fully annotated bags be automatically assembled by a simple cron job on the server.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (sentree) [inline utils/mil_tools/nodes/label_me_bag.py]: Please add a documentation string to this function. Comments would also be nice if you have time.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (sentree) [inline utils/mil_tools/nodes/label_me_bag.py]: Why no `--` before the option?

File: utils/mil_tools/nodes/label_me_bag.py
Comment (sentree) [inline utils/mil_tools/nodes/label_me_bag.py]: I don't know if extract is the right word for this. I would associate that more with the image extraction process. Maybe use inject?

File: utils/mil_tools/nodes/label_me_bag_example.yaml
Comment (sentree) [inline utils/mil_tools/nodes/label_me_bag_example.yaml]: What frequency will images in this time period be pulled?

**Review by sentree — CHANGES_REQUESTED** [review]
This is a good start on a library for interfacing with LabelMe. That said, I think this should be split up a bit. There should be a library and a CLI.

I think an RQT interface would make generating the configurations much simpler for people not familiar with the system. Having a separate library would make the RQT and CLI interfaces compatible and avoid repeated code. Basically, users would use one of the tools to extract images to the fileserver and the bag would automatically be reassembled with the labels once it is done.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (kev-the-dev) [reply utils/mil_tools/nodes/label_me_bag.py]: +1 for cron jobs

Right now without verbose it just returns one line where the last token is a percentage. You could use that.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (kev-the-dev) [reply utils/mil_tools/nodes/label_me_bag.py]: It is a positional argument not a flag.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (mattlangford) [reply utils/mil_tools/nodes/label_me_bag.py]: Not since I found `:set spell`, vims built in comment spell checker.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (mattlangford) [reply utils/mil_tools/nodes/label_me_bag.py]:
```
Use one leading underscore only for non-public methods and instance variables.

To avoid name clashes with subclasses, use two leading underscores to invoke Python's name mangling rules.
```

File: utils/mil_tools/nodes/label_me_bag.py
Comment (sentree) [reply utils/mil_tools/nodes/label_me_bag.py]: Everything is actually spelled right here, they're just not the right words. I don't really know what that's called. Thanks for the tip, though.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (sentree) [reply utils/mil_tools/nodes/label_me_bag.py]: Where did that come from? The only issue I have with one underscore is it really does nothing to make a function or variable private, it's just for looks.

**Comment by sentree** [comment]
Would it be useful to add `-n` and `--notes` arguments to the bag command that open a text file of the same name as the bag in Vim after the bag is ended?

File: utils/mil_tools/nodes/label_me_bag.py
Comment (mattlangford) [reply utils/mil_tools/nodes/label_me_bag.py]: Oh yeah, leaving out some words I see - though `corresponding` is spelled wrong

File: utils/mil_tools/nodes/label_me_bag.py
Comment (mattlangford) [reply utils/mil_tools/nodes/label_me_bag.py]:
```
We don't use the term "private" here, since no attribute is really private in Python
```
All from the PEP8 page

File: utils/mil_tools/nodes/label_me_bag_example.yaml
Comment (kev-the-dev) [reply utils/mil_tools/nodes/label_me_bag_example.yaml]: It explains in the comment on the next line that if no freq is specified, every image from this time frame will be inserted into LabelMe

File: utils/mil_tools/nodes/label_me_bag.py
Comment (kev-the-dev) [reply utils/mil_tools/nodes/label_me_bag.py]: I'm going to make it a flag though so it's more verbose.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (kev-the-dev) [reply utils/mil_tools/nodes/label_me_bag.py]: renamed to ```--inject-annotations```

File: utils/mil_tools/nodes/label_me_bag.py
Comment (kev-the-dev) [reply utils/mil_tools/nodes/label_me_bag.py]: I would prefer to have one YAML per perception task, containing all of the bags/segments used to test/train for that task. So each YAML would effectively have one maintainer, whoever is assigned that perception task. These YAMLS should be version controlled, probably in the project's git.

File: utils/mil_tools/nodes/label_me_bag.py
Comment (kev-the-dev) [reply utils/mil_tools/nodes/label_me_bag.py]: Most python libraries I look at, like rospy, use one underscore for class "private" variables/methods, so I emulated that. Python is a scripting language, so I don't think it's that important to enforce private variables when there isn't even type safety. You just got to believeeee

**Comment by kev-the-dev** [comment]
Tony's comments have been addressed. I don't think we should hold this PR up because there is not graphical interface yet. And it currently can be used as a library, with the usual ``` "if __name__ == "__main__" ``` at the bottom just providing an optional argparse CLI. The class can be used normally in a python script.

**Comment by whispercoros** [comment]
I think this is good to merge, but should we wait for a review from @DSsoto?

**Review by sentree — APPROVED** [review]
I think this is good to merge, but should we wait for a review from @DSsoto?

**Comment by kev-the-dev** [comment]
I think I fixed the unrelated build error. Should be good to go

**Pull request merged**
